drivers/sensor/icm20948/icm20948.c

```diff
@@ -0,0 +1,706 @@
+#define SELF_TEST_Z_GYRO  0x04
+#define SELF_TEST_X_ACCEL 0x0E
+#define SELF_TEST_Y_ACCEL 0x0F
+#define SELF_TEST_Z_ACCEL 0x10
+
+/* User Bank 2 */
+#define GYRO_SMPLRT_DIV    0x00
+#define GYRO_CONFIG_1      0x01
+#define GYRO_CONFIG_2      0x02
+#define ACCEL_SMPLRT_DIV_1 0x10
+#define ACCEL_SMPLRT_DIV_2 0x11
+#define ACCEL_CONFIG       0x14
+#define ACCEL_CONFIG_2     0x15
+
+/* User Bank 3 */
+#define I2C_MST_ODR_CONFIG 0x00
+#define I2C_MST_CTRL       0x01
+#define I2C_MST_DELAY_CTRL 0x02
+#define I2C_SLV0_ADDR      0x03
+#define I2C_SLV0_REG       0x04
+#define I2C_SLV0_CTRL      0x05
+#define I2C_SLV0_DO        0x06
+
+#define I2C_SLV4_ADDR 0x13
+#define I2C_SLV4_REG  0x14
+#define I2C_SLV4_CTRL 0x15
+#define I2C_SLV4_DO   0x16
+#define I2C_SLV4_DI   0x17
+/* User banks common */
+#define FIFO_EN_1     0x66
+
+#define REG_BANK_SEL 0x7F
+
+/* AK09916 */
+#define AK09916_I2C_ADDR 0x0C
+
+#define AK09916_WIA   0x01
+#define AK09916_ST1   0x10
+#define AK09916_HXL   0x11
+#define AK09916_HXH   0x12
+#define AK09916_HYL   0x13
+#define AK09916_HYH   0x14
+#define AK09916_HZL   0x15
+#define AK09916_HZH   0x16
+#define AK09916_ST2   0x18
+#define AK09916_CNTL2 0x31
+#define AK09916_CNTL3 0x32
+#define AK09916_TS1   0x33
+#define AK09916_TS2   0x34
+
+#define AK09916_DEVICE_ID 0x09
+
+#define ROOM_TEMP_OFFSET_DEG  21
+#define TEMP_SENSITIVITY_X100 33387
+#define SENS_READ_BUFF_LEN    22
+
+#define ICM20948_REG_I2C_SLV4_CTRL_VAL 0x80
+#define ICM20948_I2C_MST_STS_SLV4_DONE BIT(4)
+#define I2C_READ_FLAG                  BIT(7)
+
+#define MULT_MST_EN   BIT(7)
+#define I2C_MST_P_NSR BIT(4)
+#define I2C_MST_CLK   0x07
+
+LOG_MODULE_REGISTER(ICM20948, CONFIG_SENSOR_LOG_LEVEL);
+
+static int icm20948_bank_select(const struct device *dev, uint8_t bank)
+{
+	if (bank > 3) {
+		return -EINVAL;
+	}
+	const struct icm20948_config *cfg = dev->config;
+	const uint8_t reg_bank_mask = 0x30;
+
+	int ret = i2c_reg_update_byte_dt(&cfg->i2c, REG_BANK_SEL, reg_bank_mask, bank << 4);
+
+	if (ret) {
+		LOG_ERR("Error switching banks");
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline void icm20948_temp_c(int32_t in, int32_t *out_c, uint32_t *out_uc)
+{
+	/* Offset by 21 degrees Celsius */
+	int64_t in100 = (in * 100) + (ROOM_TEMP_OFFSET_DEG * TEMP_SENSITIVITY_X100);
+
+	/* Whole celsius */
+	*out_c = in100 / TEMP_SENSITIVITY_X100;
+
+	/* Micro celsius */
+	*out_uc = ((in100 - (*out_c) * TEMP_SENSITIVITY_X100) * INT64_C(1000000)) /
+		  TEMP_SENSITIVITY_X100;
+}
+
+static void icm20948_convert_temp(struct sensor_value *val, int16_t raw_val)
+{
+	icm20948_temp_c((int32_t)raw_val, &val->val1, &val->val2);
+}
+
+static inline void icm20948_accel_mss(int64_t sensitivity, int32_t in, int32_t *out, int32_t *u_out)
+{
+	/* Convert to micrometers/s^2 */
+	int64_t in_ms = in * SENSOR_G;
+
+	/* meters/s^2 whole values */
+	*out = in_ms / (sensitivity * 1000000LL);
+
+	/* micrometers/s^2 */
+	*u_out = (in_ms - (*out * sensitivity * 1000000LL)) / sensitivity;
```

Comment:
Doesn't this evaluate to zero?

Reply:
It doesn't evaluate to 0. The first value is the whole part of the integer division after multiplying by a million. The fractional part is obtained subtracting the integer division from the full value and then dividing this by the sensitivity as well.